### PR TITLE
EA-198: Add support for configuring (and fetching) Mother Child relat…

### PR DESCRIPTION
**Pull request by mogoodrich**
…ionships within the EMR API module

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/emrapi/EmrApiConstants.java [hunk 1]
```diff
@@ -216,5 +216,5 @@ public class EmrApiConstants {
 
     public static final String GP_USE_LEGACY_DIAGNOSIS_SERVICE = "emrapi.useLegacyDiagnosisService";
 
-    public static final String METADATA_MAPPING_MOTHER_CHILD_RELATIONSHIP_TYPE = "emrapi.motherChildRelationshipType";
+    public static final String METADATA_MAPPING_MOTHER_CHILD_RELATIONSHIP_TYPE = "emr.motherChildRelationshipType";
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/emrapi/EmrApiConstants.java:219]: All the other EMR API properties started with "emr", so I renamed this for consistency.

File: api/src/main/java/org/openmrs/module/emrapi/maternal/MaternalServiceImpl.java [hunk 2]
```diff
@@ -47,7 +47,9 @@ public List<MotherAndChild> getMothersAndChildren(MothersAndChildrenSearchCriter
 
         Map<String, Object> parameters = new HashMap<>();
         parameters.put("motherUuids", criteria.getMotherUuids());
+        parameters.put("restrictByMothers", criteria.getMotherUuids() != null);
```
Comment (mogoodrich) [inline api/src/main/java/org/openmrs/module/emrapi/maternal/MaternalServiceImpl.java:50]: see below

File: api/src/main/resources/hql/mother_child.hql [hunk 3]
```diff
@@ -6,8 +6,8 @@ from
         inner join motherChildRelationship.personA as mother
         inner join motherChildRelationship.personB as child
 where
-    ((:motherUuids) is null or mother.uuid in (:motherUuids))
-  and ((:childUuids) is null or child.uuid in (:childUuids))
+    (:restrictByMothers = false or mother.uuid in (:motherUuids))
+  and (:restrictByChildren = false or child.uuid in (:childUuids))
```
Comment (mogoodrich) [inline api/src/main/resources/hql/mother_child.hql:10]: Was stumped by this for a while... the "(:motherUuids) is null" was throwing an error when motherUuids was an array, which in-and-of-itself isn't necessarily surprising, but it was working in the tests, it was only failing when trying to use it for "real".  No idea what the difference was.  I figured I was inadvertently passing in a slightly different datatype, but debugging they looked the same.

File: omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    ) {
+        RequestContext context = RestUtil.getRequestContext(request, response, Representation.DEFAULT);
+        MothersAndChildrenSearchCriteria criteria = new MothersAndChildrenSearchCriteria();
+        criteria.setMotherUuids(StringUtils.isNotBlank(motherUuids) ? Arrays.asList(motherUuids.split(",")) : null);
+        criteria.setChildUuids(StringUtils.isNotBlank(childUuids) ? Arrays.asList(childUuids.split(",")) : null);
+        criteria.setMotherRequiredToHaveActiveVisit(requireMotherHasActiveVisit);
+        criteria.setChildRequiredToHaveActiveVisit(requireChildHasActiveVisit);
+        criteria.setChildRequiredToBeBornDuringMothersActiveVisit(requireChildBornDuringMothersActiveVisit);
+        List<MotherAndChild> motherAndChildList = maternalService.getMothersAndChildren(criteria);
+        return new NeedsPaging<>(motherAndChildList, context).toSimpleObject(new SimpleBeanConverter<MotherAndChild>());
```
Comment (mogoodrich) [inline omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java:51]: I started to create a custom converter, but at the end-of-the-day the buildt in converter seemed to give reps with reasonable defaults.  I tested full, default, rep, and a custom rep.

File: omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            @RequestParam(required = false, value = "motherUuids") String motherUuids,
+            @RequestParam(required = false, value = "childUuids") String childUuids,
+            @RequestParam(required = false, value = "requireMotherHasActiveVisit") boolean requireMotherHasActiveVisit,
+            @RequestParam(required = false, value = "requireChildHasActiveVisit") boolean requireChildHasActiveVisit,
+            @RequestParam(required = false, value = "requireChildBornDuringMothersActiveVisit") boolean requireChildBornDuringMothersActiveVisit
+    ) {
+        RequestContext context = RestUtil.getRequestContext(request, response, Representation.DEFAULT);
+        MothersAndChildrenSearchCriteria criteria = new MothersAndChildrenSearchCriteria();
+        criteria.setMotherUuids(StringUtils.isNotBlank(motherUuids) ? Arrays.asList(motherUuids.split(",")) : null);
+        criteria.setChildUuids(StringUtils.isNotBlank(childUuids) ? Arrays.asList(childUuids.split(",")) : null);
```
Comment (mogoodrich) [inline omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java:46]: Not sure if there is a built-in way to map a parameter to an array, but the few ways I tried didn't seem to work.

File: api/src/main/resources/hql/mother_child.hql (hunk 3 above)
Comment (mseaton) [reply api/src/main/resources/hql/mother_child.hql:10]: Yeah, I ran into the same thing.  I assume you saw how I ended up handling this in the other hql queries, but if not we arrived at the same solution :)

File: omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    @Autowired
+    private MaternalService maternalService;
+
+    @RequestMapping(method = RequestMethod.GET, value = "/rest/**/emrapi/maternal/mothersAndChildren")
+    @ResponseBody
+    public SimpleObject getMothersAndChildren(
+            HttpServletRequest request,
+            HttpServletResponse response,
+            @RequestParam(required = false, value = "motherUuids") String motherUuids,
+            @RequestParam(required = false, value = "childUuids") String childUuids,
```
Comment (mseaton) [inline omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java:38]: Maybe call these properties `mothers` and `children`?  Or `limitToMothers` and `limitToChildren`?

File: omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java (hunk 5 above)
Comment (mseaton) [reply omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java:46]: If it isn't supported out of the box, I'd assume it isn't considered best practice.  The way to accomplish this is typically to just pass in the request parameter multiple times.  This is one of the reasons I was thinking the singular tense maybe:  `mother=xxx&mother=yyy&mother=zzz`.  I agree that `mothers=xxx,yyy,zzz` is more concise, I just don't know if it is considered typical practice...

Comment (mogoodrich) [reply omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java:46]: Ah right, makes sense

File: api/src/main/resources/hql/mother_child.hql (hunk 3 above)
Comment (mogoodrich) [reply api/src/main/resources/hql/mother_child.hql:10]: Oops, I guess I missed that

File: omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java (hunk 6 above)
Comment (mogoodrich) [reply omod/src/main/java/org/openmrs/module/emrapi/web/controller/MothersAndChildrenController.java:38]: If best practice is not to use the comments, I would change this probably to "mother" and "child" like to specified below, let me fiddle around with this.

**Review by mogoodrich — COMMENTED** [review]
Updated based on code review, merging in, thanks @mseaton

**Pull request merged**